client/invoke.go

```diff
@@ -99,35 +94,52 @@ func Invoke(provider provider.Provider, invokeUrl string, content io.Reader, out
 		fmt.Printf(string(b) + "\n")
 	}
 
-	// for sync calls
-	if call_id, found := resp.Header[FN_CALL_ID]; found {
-		if includeCallID {
-			fmt.Fprint(os.Stderr, fmt.Sprintf("Call ID: %v\n", call_id[0]))
-		}
-		io.Copy(output, resp.Body)
-	} else {
-		// for async calls and error discovering
-		c := &callID{}
-		err = json.NewDecoder(resp.Body).Decode(c)
-		if err == nil {
-			// decode would not fail in both cases:
-			// - call id in body
-			// - error in body
-			// that's why we need to check values of attributes
-			if c.CallID != "" {
-				fmt.Fprint(os.Stderr, fmt.Sprintf("Call ID: %v\n", c.CallID))
-			} else {
-				fmt.Fprint(output, fmt.Sprintf("Error: %v\n", c.Error.Message))
-			}
-		} else {
-			return err
-		}
+	if cid, ok := resp.Header[CallIDHeader]; ok && includeCallID {
+		fmt.Fprint(output, fmt.Sprintf("Call ID: %v\n", cid[0]))
 	}
 
+	var body io.Reader = resp.Body
 	if resp.StatusCode >= 400 {
-		// TODO: parse out error message
-		return fmt.Errorf("Error calling function: status %v", resp.StatusCode)
+		// if we don't get json, we need to buffer the input so that we can
+		// display the user's function output as it was...
+		var b bytes.Buffer
+		body = io.TeeReader(resp.Body, &b)
+
+		var msg apiErr
+		err = json.NewDecoder(body).Decode(&msg)
+		if err == nil && msg.Message != "" {
+			// this is likely from fn, so unravel this...
+			return fmt.Errorf("Error invoking function. status: %v message: %v", resp.StatusCode, msg.Message)
+		}
+
+		// read anything written to buffer first, then copy out rest of body
+		body = io.MultiReader(&b, resp.Body)
 	}
 
+	// at this point, it's not an fn error, so output function output as is
+	// TODO we should give users the option to see a status code too, like call id?
```

Comment:
I'd vote yes, assuming the additional information can help them to investigate/figure out what the error is.

Reply:
think https://github.com/fnproject/cli/pull/538#issuecomment-468371086 is a good spot for this, we can keep the default plaintext to just the body for now at least

Reply:
sounds good. I just remember I was the one who added the json flag a while ago :) Might have to go through the commands again to see if there's anything I missed